Secure monitor spectre updates

jenswi-linaro opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

etienne-lms commented on core/arch/arm/sm/sm_a32.S:
```diff
@@ -314,9 +329,37 @@ UNWIND(	.fnstart)
+	read_midr r1
+	ubfx	r2, r1, #MIDR_IMPLEMENTER_SHIFT, #MIDR_IMPLEMENTER_WIDTH
+	cmp	r2, #MIDR_IMPLEMENTER_ARM
+	bne	1f
```
minor: are these 3 instruction useful since loading `sm_vect_table` as vector table is the default behaviour for not impacted cores?

jenswi-linaro replied:
Without this test the primary part number isn't defined. I'm skipping the architecture field as we're always compiling this for ARMv7.

jbech-linaro commented:
@jenswi-linaro , please add the CVE number also, so we can single out later what patches belongs to the different CVE's.

jforissier commented:
Please also add `Fixes: 3bc90f3d3ecd ("core: arm32: sm: invalidate branch predictor")`.

jenswi-linaro commented:
Rebased, squashed and tags applied.
Note that I added the same comment for `write_iciallu` as in https://github.com/OP-TEE/optee_os/pull/2055 for `thread_a32.S`.

jenswi-linaro commented:
Pushed fixes for the checkpatch error `ERROR: trailing whitespace` and warning `WARNING: Possible unwrapped commit description (prefer a maximum 75 chars per line)`

jenswi-linaro commented:
One checkpatch warning for a long line remains, I don't want to break the url.

etienne-lms commented:
Late comment:
The change states that c-a15 way is `For that CPU, set ACTLR[0] to 1 during early processor initialisation, and invalidate the branch predictor by performing an ICIALLU instruction.`, but we do not enable this bit at core boot. 
This must be done per platform as there is no generic c-a15 core init routine. Impacted platforms are sunxi, ti-dra7xx and qemu_virt.

jenswi-linaro commented:
Good point, I'll make another PR as this PR and https://github.com/OP-TEE/optee_os/pull/2055 will need this.

jenswi-linaro commented:
Needs https://github.com/OP-TEE/optee_os/pull/2071

jenswi-linaro commented:
Rebased on master
